File: python/invaders_game/player.py
"""自機の入力処理・移動・描画を担当するモジュール。"""

import pygame


class Player:
    """左右移動する自機。"""

    WIDTH = 48
    HEIGHT = 28
    BASE_SPEED = 6

    def __init__(self, screen_width: int, ground_y: int) -> None:
        """
        Args:
            screen_width: 画面幅
            ground_y: 自機の下端が揃うY座標（地面ライン）
        """
        self.screen_width = screen_width
        self.rect = pygame.Rect(0, 0, self.WIDTH, self.HEIGHT)
        self.rect.midbottom = (screen_width // 2, ground_y)

    def reset_position(self) -> None:
        """残機減少後などに中央へ戻す。"""
        self.rect.midbottom = (self.screen_width // 2, self.rect.bottom)

    def update(self, keys: pygame.key.ScancodeWrapper) -> None:
        """←→キーで横移動。画面内にクランプする。"""
        dx = 0
        if keys[pygame.K_LEFT]:
            dx -= self.BASE_SPEED
        if keys[pygame.K_RIGHT]:
            dx += self.BASE_SPEED
        self.rect.x += dx
        if self.rect.left < 0:
            self.rect.left = 0
        if self.rect.right > self.screen_width:
            self.rect.right = self.screen_width

    def draw(self, surface: pygame.Surface) -> None:
        """自機を描画する（簡易キャノン形状）。"""
        body = self.rect.inflate(-8, -6)
        pygame.draw.rect(surface, (180, 220, 255), body, border_radius=3)
        tip_w = 10
        tip = pygame.Rect(0, 0, tip_w, 8)
        tip.midtop = (self.rect.centerx, self.rect.top)
        pygame.draw.rect(surface, (140, 200, 255), tip)

File: java/invaders_game/Player.java
/**
 * ファイル名：Player.java
 * 説明：自機の移動・位置リセット・描画を担当する。
 * 作成日：2026-04-10
 * 作者：takashioikawa
 */

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.Rectangle;

/** 左右移動する自機。 */
public class Player {

    public static final int WIDTH = 48;
    public static final int HEIGHT = 28;
    /** 1フレームあたりの横移動量 */
    public static final int BASE_SPEED = 6;

    private final int screenWidth;
    private final Rectangle rect;

    public Player(int screenWidth, int groundY) {
        this.screenWidth = screenWidth;
        this.rect = new Rectangle(0, 0, WIDTH, HEIGHT);
        rect.x = (screenWidth - WIDTH) / 2;
        rect.y = groundY - HEIGHT;
    }

    public Rectangle getBounds() {
        return rect;
    }

    /** 残機減少後などに画面中央へ戻す。 */
    public void resetPosition() {
        rect.x = (screenWidth - WIDTH) / 2;
    }

    /**
     * ←→キー入力に応じて横移動し、画面内に収める。
     *
     * @param leftPressed  左キーが押されているか
     * @param rightPressed 右キーが押されているか
     */
    public void update(boolean leftPressed, boolean rightPressed) {
        int dx = 0;
        if (leftPressed) {
            dx -= BASE_SPEED;
        }
        if (rightPressed) {
            dx += BASE_SPEED;
        }
        rect.x += dx;
        if (rect.x < 0) {
            rect.x = 0;
        }
        if (rect.x + WIDTH > screenWidth) {
            rect.x = screenWidth - WIDTH;
        }
    }

    /** 自機を描画する（簡易キャノン形状）。 */
    public void draw(Graphics2D g2) {
        int bodyX = rect.x + 4;
        int bodyY = rect.y + 3;
        int bodyW = WIDTH - 8;
        int bodyH = HEIGHT - 6;
        g2.setColor(new Color(180, 220, 255));
        g2.fillRoundRect(bodyX, bodyY, bodyW, bodyH, 6, 6);
        int tipW = 10;
        int tipH = 8;
        int tipX = rect.x + (WIDTH - tipW) / 2;
        g2.setColor(new Color(140, 200, 255));
        g2.fillRect(tipX, rect.y, tipW, tipH);
    }
}
